component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java

```diff
@@ -302,28 +319,36 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
         long generatedTime = (long) context.getProperty(EmailOTPAuthenticatorConstants.OTP_GENERATED_TIME);
         boolean isExpired = isExpired(generatedTime, context);
 
-        AuthenticatedUser authenticatedUser = getAuthenticatedUser(context);
-        if (authenticatedUser == null) {
-            String errorMessage = "Could not find an Authenticated user in the context.";
-            throw new AuthenticationFailedException(errorMessage);
-        }
-
-        if (userToken.equals(contextToken) && !isExpired) {
-            processValidUserToken(context, authenticatedUser);
-        } else if (isBackupCodeEnabled(context)) {
-            validateWithBackUpCodes(context, userToken, authenticatedUser);
-        } else {
-            if (isExpired) {
-                if (log.isDebugEnabled()) {
-                    log.debug("Given otp code is expired.");
-                }
-                throw new AuthenticationFailedException("Code expired");
+        try {
+            if (userToken.equals(contextToken) && !isExpired) {
+                processValidUserToken(context, authenticatedUser);
+            } else if (isLocalUser && isBackupCodeEnabled(context)) {
+                validateWithBackUpCodes(context, userToken, authenticatedUser);
             } else {
-                if (log.isDebugEnabled()) {
-                    log.debug("Given otp code is mismatch");
+                if (isExpired) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Given otp code is expired.");
+                    }
+                    throw new AuthenticationFailedException("Code expired");
+                } else {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Given otp code is mismatch");
+                    }
+                    throw new AuthenticationFailedException("Code mismatch");
                 }
-                throw new AuthenticationFailedException("Code mismatch");
             }
+        } catch (AuthenticationFailedException e){
+            // Check whether account locking enabled for Email OTP to keep backward compatibility
+            if (isLocalUser && EmailOTPUtils.isAccountLockingEnabledForEmailOtp(context)) {
+                handleOtpVerificationFail(authenticatedUser);
+            }
+            throw e;
+        }
+
+        // It reached here means the authentication was successful
+        // Check whether account locking enabled for Email OTP to keep backward compatibility
+        if (isLocalUser && EmailOTPUtils.isAccountLockingEnabledForEmailOtp(context)) {
+            resetOtpFailedAttempts(authenticatedUser);
```

Comment:
cant we derive whether the user is local or not, within the `resetOtpFailedAttempts` method?